src/main/resources/db/changelog/20240802101300-update-anon-cert-table-cert-column.xml

```diff
@@ -0,0 +1,13 @@
+<databaseChangeLog
+        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+        http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.19.xsd">
+
+    <changeSet id="20240802101300-1" author="jalbrech">
```

Comment:
Updated the change set to use the `BLOB(len)` to avoid using the aliases and translations and also added the limitation to only run the changeset on MariaDb/MySQL.

Reply:
> so we may be best off defining the column like that (if the length definitions line up) rather than relying on aliases and translations we don't control.

By this do you mean to set the max length of the mariadb/mysql column size to the max column size of the Postgres bytea column size (1 GB)?

Reply:
Nah, I was just meaning that HSQLDB allows defining a blob column length with shortcuts like `1G` or `64k`. I don't recall what MySQL/MariaDB does, and I know postgresql doesn't follow the standard at all. But at least for HSQLDB and the MySQL variants, it may be possible to use one definition to get the same column explicitly.

But yeah, we should probably try to line those up where we can as well. In terms of length, 1gb is probably more than safe. If we can't make our DB definitions line up nicely, we should at least document our largest possible cert somewhere so it's not a surprise when it blows up in one and not another.

Reply:
Just checked: looks like 1gb is our absolute max due to PostgreSQL limits on bytea columns. Interestingly enough, that's also the default size of a BLOB column when no length is defined.

So really the outlier here is, as per usual, MySQL with the 65k size on BLOB. If we change the definition in datatypes.xml to `BLOB(1073741824)` (barring a nicer syntax for that definition) for MySQL/MariaDB/HSQLDB, we should get the following:

- PostgreSQL: bytea, 1GB
- MySQL/MariaDB: LONGBLOB
- HSQLDB: BLOB(1G)

MySQL is still going to be weird and let us go to 4gb, but I think that'll require some malicious intent to actually hit so it's *probably* fine until I have to eat my words in a year or two when someone manages to hit it.

Reply:
Alternatively, we could define the three DBs differently and not need to worry about one explicit syntax for MySQL/MariaDB and HSQLDB.

Reply:
I updated the datatypes.xml to `BLOB(1073741824)` for MySQL/MariaDB/HSQLDB, but I'm still getting a `BLOB` and not a `LONGBLOB` datatype for the cert column in MySQL.

Reply:
That's not what it should be doing. MySQL/MariaDB state that when you provide a length, it then picks the type that will fit that size -- which for 1gb should be a LONGBLOB type.

> An optional length M can be given for this type. If this is done, MariaDB creates the column as the smallest BLOB type large enough to hold values M bytes long.

In light of that, we're best off having three cert.type definitions: one for HSQLDB that  uses BLOB, optionally with the 1gb length specifier (though that's also the default so I'll leave that preference up to y'all), one for MySQL/MariaDB that explicitly defines it as a LONGBLOB, and then one for PostgreSQL that defines it as a bytea.

Note that with Liquibase, you can't *redefine* a type once it's defined, so the order in which those are defined matters if there's any overlap. I don't think there is in this case, so this is mostly just an FYI on how that owrks.

Reply:
Isn't the issue with using `BLOB(1073741824)`  as the cert.type value for MySQL/MariaDB/HSQLDB that the value should be a generic Liquibase datatype? Then Liquibase converts the generic datatype to whatever datatype that is used for the underlying database. So, the Liquibase abstraction could be mapping that `BLOB(1073741824)` to a MySQL Blob instead of LongBlob.

Like if I alter the MySQL table using the following statement, the column is updated to Longblob,  but using Liquibase results in a Blob.

`ALTER TABLE cp_anonymous_certificates MODIFY cert BLOB(1073741824);`